znnl/analysis/loss_ntk_calculation.py

```diff
@@ -0,0 +1,241 @@
+
+        dataset : dict
+                The dataset for which to calculate the loss NTK.
+                The dictionary should contain the keys "inputs" and "targets".
+        """
+
+        # Set the attributes
+        self.ntk_batch_size = model.ntk_batch_size
+        self.store_on_device = model.store_on_device
+        self.trace_axes = model.trace_axes
+        self.input_shape = dataset["inputs"].shape
+        self.input_dimension = int(np.prod(np.array(self.input_shape[1:])))
+        self.target_shape = dataset["targets"].shape
+        self.metric_fn = metric_fn
+        self.apply_fn = model._ntk_apply_fn
+
+        # Prepare NTK calculation
+        empirical_ntk = nt.batch(
+            nt.empirical_ntk_fn(
+                f=self._function_for_loss_ntk,
+                trace_axes=(),
+                vmap_axes=0,
+            ),
+            batch_size=self.ntk_batch_size,
+            store_on_device=self.store_on_device,
+        )
+        self.empirical_ntk_jit = jax.jit(empirical_ntk)
+
+    @staticmethod
+    def _reshape_dataset(dataset):
+        """
+        Helper function to reshape the dataset for the Loss NTK calculation.
+
+        Parameters
+        ----------
+        dataset : dict
+                The dataset to be reshaped.
+                Should contain the keys "inputs" and "targets".
+
+        Returns
+        -------
+        reshaped_dataset : np.ndarray
+                The reshaped dataset.
+        """
+        return np.concatenate(
+            (
+                dataset["inputs"].reshape(dataset["inputs"].shape[0], -1),
+                dataset["targets"].reshape(dataset["targets"].shape[0], -1),
+            ),
+            axis=1,
+        )
+
+    @staticmethod
+    def _unshape_data(
+        datapoint: np.ndarray,
+        input_dimension: int,
+        input_shape: tuple,
+        target_shape: tuple,
+        batch_length: int,
+    ):
+        """
+        Helper function to unshape the data for the subloss calculation.
+
+        Parameters
+        ----------
+        datapoint : np.ndarray
+                The datapoint to be unshaped.
+        input_dimension : int
+                The total dimension of the input, i.e. the product of its shape.
+        input_shape : tuple
+                The shape of the original input.
+        target_shape : tuple
+                The shape of the original target.
+        batch_length : int
+                The length of the batch, i.e. the number of actual datapoints
+                inside the datapoint variable
+
+        Returns
+        -------
+        input: np.ndarray
+                The unshaped input.
+        target: np.ndarray
+                The unshaped target.
+        """
+        return datapoint[:, :input_dimension].reshape(
+            batch_length, *input_shape[1:]
+        ), datapoint[:, input_dimension:].reshape(batch_length, *target_shape[1:])
+
+    def _function_for_loss_ntk(self, params, datapoint) -> float:
```

Comment:
I would prefer different naming here. Is it an apply function on flattened data, a loss function. What do you mean by subloss? Loss between two data points is just loss. Function for loss ntk could be anything.